File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_CREATEACCOUNT2.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_CREATEACCOUNT2 : Message
    {
        public SID_CREATEACCOUNT2()
        {
            Id = (byte)MessageIds.SID_CREATEACCOUNT2;
            Buffer = new byte[0];
        }

        public SID_CREATEACCOUNT2(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_CREATEACCOUNT2;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

                        if (Buffer.Length < 21)
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 21 bytes");

                        /**
                         * (UINT32) [5] Password hash
                         * (STRING) Username
                         */

                        using var m = new MemoryStream(Buffer);
                        using var r = new BinaryReader(m);

                        var passwordHash = r.ReadBytes(20);
                        var username = r.ReadString();

                        Account.CreateStatus status = Account.TryCreate(username, passwordHash, out var _);

                        return new SID_CREATEACCOUNT2().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {
                            { "status", status } // info key optional
                        }));
                    }
                case MessageDirection.ServerToClient:
                    {
                        /**
                         * (UINT32) Status
                         * (STRING) Account name suggestion
                         */

                        string info = context.Arguments.ContainsKey("info") ? (string)context.Arguments["info"] : "";

                        Buffer = new byte[5 + Encoding.UTF8.GetByteCount(info)];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write((UInt32)(Account.CreateStatus)context.Arguments["status"]);
                        w.Write(info);

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_CREATEACCOUNT2.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_CREATEACCOUNT2
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_CREATEACCOUNT2)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_CREATEACCOUNT2)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Dim __ As Account = Nothing

            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    If Buffer.Length < 21 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 21 bytes")

                    Dim passwordHash() As Byte, username As String
                    Using m = New MemoryStream(Buffer)
                        Using r = New BinaryReader(m)
                            passwordHash = r.ReadBytes(20)
                            username = r.ReadString()
                        End Using
                    End Using

                    Dim status As Account.CreateStatus = Account.TryCreate(username, passwordHash, __)
                    Return New SID_CREATEACCOUNT2().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object) From {
                        {"status", status}
                    }))
                Case MessageDirection.ServerToClient
                    Dim info As String = If(context.Arguments.ContainsKey("info"), CStr(context.Arguments("info")), "")
                    Buffer = New Byte(5 + Encoding.UTF8.GetByteCount(info) - 1) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(CUInt(CType(context.Arguments("status"), Account.CreateStatus)))
                            w.Write(info)
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
